airbyte-cdk/python/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py

```diff
@@ -350,7 +360,26 @@ def _create_nested_component(self, model, model_field: str, model_value: Any, co
         model_type = self.TYPE_NAME_TO_MODEL.get(type_name, None)
         if model_type:
             parsed_model = model_type.parse_obj(model_value)
-            return self._create_component_from_model(model=parsed_model, config=config)
+            try:
+                # To improve usability of the language, certain fields are shared between components. This can come in the form of
+                # a parent component passing some of its fields to a child component or the parent extracting fields from other child
+                # components and passing it to others. One example is the DefaultPaginator referencing the HttpRequester url_base
+                # while constructing a SimpleRetriever. However, custom components don't support this behavior because they are created
+                # generically in create_custom_component(). This block allows developers to specify extra arguments in $parameters that
+                # are needed by a component and could not be shared.
+                model_constructor = self.PYDANTIC_MODEL_TO_CONSTRUCTOR.get(parsed_model.__class__)
+                constructor_kwargs = inspect.getfullargspec(model_constructor).kwonlyargs
+                model_parameters = model_value.get("$parameters", {})
+                matching_parameters = {kwarg: model_parameters[kwarg] for kwarg in constructor_kwargs if kwarg in model_parameters}
+                return self._create_component_from_model(model=parsed_model, config=config, **matching_parameters)
+            except TypeError as error:
+                missing_parameters = self._extract_missing_parameters(error)
+                if missing_parameters:
+                    raise ValueError(
+                        f"Error creating component '{type_name}' with parent custom component {model.class_name}: Please provide "
+                        + ", ".join((f"{type_name}.$parameters.{parameter}" for parameter in missing_parameters))
+                    )
+                raise
```

Comment:
In the case where the error is not related to missing parameters, should we provide more guidance to the user as to in which context it's failing? I think reusing the message you had makes sense. Something like: `raise TypeError("Error creating component '{type_name}' with parent custom component {model.class_name}: {e}")`

Reply:
good point! updated